### PR TITLE
BP-1091 - Customize WithErrorLogs interceptor with debug/info level options

### PR DESCRIPTION
**Pull request by sp-francisco-moreno**

### CODE REVIEW

File: grpcx/server.go [hunk 1]
```diff
@@ -56,22 +58,71 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
 
 // WithErrorLogs returns a gRPC interceptor for unary calls that instrument requests
 // with logs for the errors.
-func WithErrorLogs(l logx.Logger) grpc.UnaryServerInterceptor {
+func WithErrorLogs(l logx.Logger, options ...func(*withErrorLogsOptions)) grpc.UnaryServerInterceptor {
+	var logOptions = &withErrorLogsOptions{
+		debugLevelCodes: []codes.Code{},
+		infoLevelCodes:  []codes.Code{codes.Canceled, codes.Unknown, codes.InvalidArgument, codes.DeadlineExceeded, codes.NotFound, codes.AlreadyExists, codes.PermissionDenied, codes.ResourceExhausted, codes.FailedPrecondition, codes.Aborted, codes.OutOfRange, codes.Unimplemented, codes.Internal, codes.Unavailable, codes.DataLoss, codes.Unauthenticated},
+	}
```
Comment (sp-ruben-simon) [inline grpcx/server.go:65]: what if by default all the errors are logged as Info and we add an option to configure errors as Debug and another option to don't log it?

Comment (sp-jordi-forns) [reply grpcx/server.go:65]: +1

File: grpcx/server.go [hunk 2]
```diff
@@ -56,22 +58,71 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
 
 // WithErrorLogs returns a gRPC interceptor for unary calls that instrument requests
 // with logs for the errors.
-func WithErrorLogs(l logx.Logger) grpc.UnaryServerInterceptor {
+func WithErrorLogs(l logx.Logger, options ...func(*withErrorLogsOptions)) grpc.UnaryServerInterceptor {
```
Comment (sp-jordi-forns) [inline grpcx/server.go:61]: the argument in the options func is not public, so callers won't be able to use them.

see an example impl here: https://github.com/socialpoint/jarvis-manager/blob/76351bbd7ab2b3d14533adf4f1d606f1b924eb67/pkg/apps/github/github.go#L69-L89

File: grpcx/server.go [hunk 3]
```diff
@@ -56,22 +58,71 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
...
+func inArray(needle codes.Code, haystack []codes.Code) bool {
+	for _, elem := range haystack {
+		if elem == needle {
+			return true
+		}
+	}
+	return false
+}
+
+func SetDebugLevelCodes(codes []codes.Code) func(*withErrorLogsOptions) {
```
Comment (sp-jordi-forns) [inline grpcx/server.go:118]: option funcs typically start with the word `With`, so this one would be `WithDebugLevelCodes`.

File: grpcx/server.go [hunk 4]
```diff
@@ -56,22 +58,71 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
...
 				{Key: "ctx_response_content", Value: string(respMsg)},
-				{Key: "ctx_response_error", Value: err.Error()},
+				{Key: "ctx_response_error_code", Value: errCode},
+				{Key: "ctx_response_error_message", Value: err.Error()},
+			}
+			if logOptions.hasDebugLevel(errCode) {
+				l.Debug("gRPC Error", fields...)
+			}
+			if logOptions.hasInfoLevel(errCode) {
+				l.Info("gRPC Error", fields...)
```
Comment (sp-jordi-forns) [inline grpcx/server.go:88]: options are just DTOs.
these responsibilities belong to the `WithErrorLogs` domain. although they could just be plain funcs.

**Review by sp-jordi-forns — CHANGES_REQUESTED** [review]

File: grpcx/server.go (hunk 3 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go]: 👍  changed!

File: grpcx/server.go (hunk 2 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go]: 👍   good catch! Changed!

File: grpcx/server.go (hunk 4 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go:88]: Ok, changed

File: grpcx/server.go (hunk 1 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go]: Ok, I've added WithoutLevelCodes option, where you can exclude the codes you don't want.

File: grpcx/server.go [hunk 5]
```diff
@@ -56,22 +58,78 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
 
 // WithErrorLogs returns a gRPC interceptor for unary calls that instrument requests
 // with logs for the errors.
-func WithErrorLogs(l logx.Logger) grpc.UnaryServerInterceptor {
+func WithErrorLogs(l logx.Logger, options ...func(*WithErrorLogsOptions)) grpc.UnaryServerInterceptor {
+	var logOptions = &WithErrorLogsOptions{
+		debugLevelCodes: []codes.Code{},
+		infoLevelCodes:  []codes.Code{codes.Canceled, codes.Unknown, codes.InvalidArgument, codes.DeadlineExceeded, codes.NotFound, codes.AlreadyExists, codes.PermissionDenied, codes.ResourceExhausted, codes.FailedPrecondition, codes.Aborted, codes.OutOfRange, codes.Unimplemented, codes.Internal, codes.Unavailable, codes.DataLoss, codes.Unauthenticated},
```
Comment (sp-jordi-forns) [inline grpcx/server.go:64]: i don't think you need to have a list with all the possible codes. 
also what happens if they add a new one? you'll have to change this and that's a pita.
you should only need one list with the codes with debug level and the codes to discard.

File: grpcx/server.go [hunk 6]
```diff
@@ -56,22 +58,78 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
...
 		}
 
 		return resp, err
 	}
 }
+
+type WithErrorLogsOptions struct {
+	debugLevelCodes []codes.Code
+	infoLevelCodes  []codes.Code
+}
```
Comment (sp-jordi-forns) [inline grpcx/server.go:99]: this type doesn't need to be public. make it private and define a type for `func(*WithErrorLogsOptions)` as in the examples.

File: grpcx/server.go [hunk 7]
```diff
@@ -56,22 +58,78 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
...
+		logsConfig.infoLevelCodes = codes
+	}
+}
+
+func WithoutLevelCodes(codes []codes.Code) func(*WithErrorLogsOptions) {
+	return func(logsConfig *WithErrorLogsOptions) {
+		removeCodesFromList(codes, logsConfig.debugLevelCodes)
+		removeCodesFromList(codes, logsConfig.infoLevelCodes)
+	}
+}
```
Comment (sp-jordi-forns) [inline grpcx/server.go:127]: this should go into a discard list, not mess with the others.

File: grpcx/server.go [hunk 8]
```diff
@@ -56,22 +58,78 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
...
 		return resp, err
 	}
 }
+
+type WithErrorLogsOptions struct {
+	debugLevelCodes []codes.Code
+	infoLevelCodes  []codes.Code
+}
+
+func inArray(needle codes.Code, haystack []codes.Code) bool {
```
Comment (sp-jordi-forns) [inline grpcx/server.go:101]: this naming is leaking php like hell

**Review by sp-jordi-forns — CHANGES_REQUESTED** [review]

File: grpcx/server.go (hunk 5 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go]: Ok, replaced by debugLevelCodes & discardedCodes

File: grpcx/server.go (hunk 6 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go:103]: Ok, now it's private again, and WithErrorLogsOption it's public

File: grpcx/server.go (hunk 7 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go]: 👍  changed! Now you add the codes to the discardedCodes list.

File: grpcx/server.go (hunk 8 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go]: 😅  you're right! changed to `inCodeList`

File: grpcx/server.go [hunk 9]
```diff
@@ -56,22 +58,67 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
 
 // WithErrorLogs returns a gRPC interceptor for unary calls that instrument requests
 // with logs for the errors.
-func WithErrorLogs(l logx.Logger) grpc.UnaryServerInterceptor {
+func WithErrorLogs(l logx.Logger, options ...WithErrorLogsOption) grpc.UnaryServerInterceptor {
```
Comment (sp-jordi-forns) [inline grpcx/server.go:61]: s/WithErrorLogsOption/ErrorLogsOption

File: grpcx/server.go [hunk 10]
```diff
@@ -56,22 +58,67 @@ func WithRequestResponseLogs(l logx.Logger) grpc.UnaryServerInterceptor {
...
+	var logOptions = &withErrorLogsOptions{
+		debugLevelCodes: []codes.Code{},
+		discardedCodes:  []codes.Code{},
```
Comment (sp-jordi-forns) [inline grpcx/server.go:64]: i don't even think you need to initialize them.

**Review by sp-jordi-forns — CHANGES_REQUESTED** [review]
almost there :)

File: grpcx/server.go (hunk 9 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go]: done!

File: grpcx/server.go (hunk 10 above)
Comment (sp-francisco-moreno) [reply grpcx/server.go]: you're right!

**Review by sp-jordi-forns — APPROVED** [review]

**Pull request merged**